### PR TITLE
Implementation of VaultGemma Fine Tuning with Differential Privacy and Inference

### PR DESCRIPTION
**Pull request by RubensZimbres**
# Add VaultGemma Fine-tuning with Differential Privacy and Inference

## Overview
This PR adds a complete pipeline for privacy-preserving fine-tuning and inference of VaultGemma 1B on medical data using LoRA adapters and differential privacy via Opacus.

## What's Added
- Complete training pipeline with 4-bit quantization, LoRA, and differential privacy
- Inference code for loading and running fine-tuned models
- Comprehensive README with setup instructions and usage examples
- Notebook: `VaultGemma_FineTuning_Inference_Huggingface.ipynb`

## Key Features
- 4-bit quantization using BitsAndBytes NF4 for memory efficiency
- LoRA fine-tuning targeting all projection layers with r=8, alpha=16
- Differential privacy with configurable epsilon and delta budgets
- Prompt masking to train only on response tokens
- Automatic checkpointing based on loss thresholds
- Cosine learning rate schedule with warmup

### Training Configuration
- Model: VaultGemma 1B (4-bit quantized)
- Dataset: Medical Meadow Medical Flashcards (1000 samples)
- Privacy Budget: ε=8.0, δ=1e-5
- Batch Size: 1 with gradient accumulation of 8
- Learning Rate: 2e-5 with cosine schedule
- Train/Validation Split: 90/10

### Privacy Guarantees
The implementation provides (ε, δ)-differential privacy guarantees through gradient clipping (max norm: 1.0) and automatic privacy accounting via Opacus.

### Inference
Includes simple inference functions with adjustable generation parameters (temperature, top_p, max_new_tokens) and support for single or batch processing.

## Files Changed
- `VaultGemma_FineTuning_Inference_Huggingface.ipynb` - Main training and inference notebook
- `README.md` - Documentation

### CODE REVIEW

**Comment by bebechien** [comment]
Hi @RubensZimbres 
Thanks for your contribution.
Since VaultGemma is research focused model. I think [Research](https://github.com/google-gemini/gemma-cookbook/tree/main/Research) folder is the correct place for this notebook.

So, could you modify it with the following points?

1. move VaultGemma/VaultGemma_FineTuning_Inference_Huggingface.ipynb -> Reserach/[VaultGemma]FineTuning_with_Huggingface.ipynb
2. run nbfmt script

```
$ python3 -m pip install -U --user git+https://github.com/tensorflow/docs
$ python3 -m tensorflow_docs.tools.nbfmt notebook.ipynb
```

**Comment by RubensZimbres** [comment]
Done, @bebechien , as well as gemini-code-assist issues addressed.

**Review by bebechien — APPROVED** [review]
lgtm!
